File: 304-range-sum-query-2d-immutable/range-sum-query-2d-immutable.py
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        rows , cols = len(matrix) , len(matrix[0])

        self.NewMatrix = [[0] * (cols + 1) for _ in range(rows + 1)]

        for r in range(rows):
            prefix = 0
            for c in range(cols):
                prefix += matrix [r][c]

                above = self.NewMatrix[r][c + 1]
                self.NewMatrix[r + 1][c + 1]= prefix + above
        

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        row1 += 1
        col1 += 1
        row2 += 1
        col2 += 1

        bottomright = self.NewMatrix[row2][col2]
        topleft = self.NewMatrix[row1-1][col1-1]
        left = self.NewMatrix[row2][col1-1]
        above = self.NewMatrix[row1-1][col2]

        return bottomright - left - above + topleft

        

File: 304-range-sum-query-2d-immutable/range-sum-query-2d-immutable.java
class NumMatrix {
    int prefix [][];
    public NumMatrix(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        prefix = new int[rows+1][cols+1];

        for( int r = 0 ; r < rows ; r++)
        {
            for(int c = 0 ; c < cols ; c++)
            {
                prefix[r + 1][c + 1] = matrix[r][c] + 
                prefix[r][c + 1] +
                prefix[r + 1][c] -
                prefix[r][c] ;
            }
        }
        
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {

        return prefix[row2 + 1 ][col2 + 1] -
        prefix[row1][col2 + 1] -
        prefix[row2 + 1][col1] +
        prefix[row1][col1] ;


        
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */